src/compiler/ast-node-conv.bas

```diff
@@ -472,6 +484,41 @@ function astNewCONV _
 	n->l = l
 	n->cast.doconv = doconv
 	n->cast.do_convfd2fs = FALSE
+	n->cast.convconst = FALSE
+
+	'' Discarding/changing const qualifier bits ?
+	'' TODO: reuse symbCheckConstAssign()?
```

Comment:
yea, I think re-using symbCheckConstAssign and astPtrCheck (see astNewASSIGN/hCheckConstAndPointerOps) would be useful. The code below seems to be effectively the same as symbCheckConstAssign, and astPtrCheck uses the overload resolution (typeCalcMatch()) which will also check procedure pointer parameter types recursively.

```
sub sub_const(byref i as const integer)
	print i
end sub

sub sub_noconst(byref i as integer)
	i = 123
	print i
end sub

type type_const as sub(byref as const integer)
type type_noconst as sub(byref as integer)

dim ptr_const as type_const
dim ptr_noconst as type_noconst

ptr_const   = @sub_const   '' safe, same type
ptr_const   = @sub_noconst '' unsafe, because when calling through the ptr the param appears const, but the sub actually modifies it
ptr_noconst = @sub_const   '' safe, ptr allows more than the sub will do
ptr_noconst = @sub_noconst '' safe, same type

print cptr(sub(byref as const integer), @sub_const) '' safe
print cptr(sub(byref as const integer), @sub_noconst) '' unsafe, currently no warning even with -w constness
print cptr(sub(byref as integer), @sub_const) '' safe
print cptr(sub(byref as integer), @sub_noconst) '' safe
```

Reply:
I see.  Thanks.  So, need to go deeper here.  I will study to reuse the existing checking functions.  I have only been looking for places where might be manipulating the ast type directly.

A few months ago with https://sourceforge.net/p/fbc/bugs/880/ I was looking at typeCalcMatch() and related symb-proc.bas functions.  I started experimenting with:
* scoring the same const vs different const matches separately by increasing the range of the score 0 to 4*FB_DATATYPES
* counting the exact_matches and the const_matches separately
* making AST_OPOPT_NOCOERCION as 2 flags, one for left hand side, one for right hand side (though, I don't know if this helped and at the moment I can't remember why I was trying this)
I abandoned that effort, realizing I needed to study the code paths a lot more.  I dunno, I don't want to make this change too difficult, but maybe worth connecting the two ends of this experiment/effort.